### PR TITLE
Use the composite prober with the broker

### LINKED ISSUE #3187: Use the composite prober for the broker
**Problem**
Once https://github.com/knative-sandbox/eventing-kafka-broker/issues/3176 is complete, we should use the composite prober to probe the broker, instead of the async prober we are currently using. 

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)** 
Which persona is this feature for?

**Exit Criteria**
The broker probes with the composite prober instead of the async prober, and resolves properly for all transport encryption flags.

**Time Estimate (optional):** 
How many developer-days do you think this may take to resolve? 1

**Additional context (optional)**
Add any other context about the feature request here.
- https://github.com/knative-sandbox/eventing-kafka-broker/pull/3178
- https://github.com/knative-sandbox/eventing-kafka-broker/issues/3141

### PR DESCRIPTION
**Pull request by Cali0707**
Part of #3141 
Fixes #3187 

## Proposed Changes

- Update the broker controller to use the prober with tls if the transport encryption flag is set to strict.

**Release Note**

```release-note
The broker now uses a tls-enabled prober if the transport encryption flag is set to strict.
```

### CODE REVIEW

File: control-plane/pkg/reconciler/broker/broker.go [hunk 1]
```diff
@@ -683,7 +685,7 @@ func finalizerSecret(object metav1.Object) string {
 }
 
 func (r *Reconciler) getCaCerts() (string, error) {
-	secret, err := r.SecretLister.Secrets(r.SystemNamespace).Get(brokerIngressTLSSecretName)
+	secret, err := r.SecretLister.Secrets(eventingNamespace).Get(brokerIngressTLSSecretName)
```
Comment (Cali0707) [inline control-plane/pkg/reconciler/broker/broker.go:688]: @pierDipi I had to change this because we are installing the secrets into the `knative-eventing` namespace currently (https://github.com/knative-sandbox/eventing-kafka-broker/blob/main/data-plane/config/broker-tls/broker-ingress-tls-certificate.yaml#L19)

**Comment by Cali0707** [comment]
@pierDipi any ideas on if/how we should test that this is working?

**Review by aliok — COMMENTED** [review]
LGTM

File: control-plane/pkg/reconciler/broker/broker.go (hunk 1 above)
Comment (pierDipi) [reply control-plane/pkg/reconciler/broker/broker.go:700]: I'd leave it to be SystemNamespace, if namespaced broker fails, we should disable TLS for namespaced broker for now until we discuss how (and if) do we want to solve TLS for namespaced broker

File: control-plane/pkg/reconciler/broker/namespaced_controller.go [hunk 2]
```diff
@@ -115,7 +116,18 @@ func NewNamespacedController(ctx context.Context, watcher configmap.Watcher, env
...
+	caCerts, err := reconciler.getCaCerts()
+	if err != nil && (features.IsStrictTransportEncryption() || features.IsPermissiveTransportEncryption()) {
+		// We only need to warn here as the broker won't reconcile properly without the proper certs because the prober won't succeed
+		logger.Warn("Failed to get CA certs when at least one address uses TLS", zap.Error(err))
+	}
+
+	reconciler.Prober, err = prober.NewComposite(ctx, env.IngressPodPort, env.IngressPodTlsPort, reconciler.IPsLister.List, impl.EnqueueKey, &caCerts)
+	if err != nil {
+		logger.Fatal("Failed to created prober", zap.Error(err))
+	}
```
Comment (pierDipi) [inline control-plane/pkg/reconciler/broker/namespaced_controller.go:130]: Currently, TLS is not supported on namespaced broker and we haven't defined how (and if) we want to implement TLS for it (there are a few reasons like data plane can't mount secrets cross namespace, etc)

I'd not do this here since we need to discuss those issues/problems for namespaced broker

File: control-plane/pkg/reconciler/broker/namespaced_controller.go [hunk 3]
```diff
@@ -212,3 +224,15 @@ func NewNamespacedController(ctx context.Context, watcher configmap.Watcher, env
...
+	secret, err := r.SecretLister.Secrets(r.SystemNamespace).Get(brokerIngressTLSSecretName)
+	if err != nil {
+		return "", fmt.Errorf("failed to get CA certs from %s/%s: %w", r.SystemNamespace, brokerIngressTLSSecretName, err)
+	}
+	caCerts, ok := secret.Data[caCertsSecretKey]
+	if !ok {
+		return "", fmt.Errorf("failed to get CA certs from %s/%s: missing %s key", r.SystemNamespace, brokerIngressTLSSecretName, caCertsSecretKey)
+	}
+	return string(caCerts), nil
+}
```
Comment (pierDipi) [inline control-plane/pkg/reconciler/broker/namespaced_controller.go:238]: Same as above, I'd leave out the TLS for namespaced broker until we have a design that would work for it

File: control-plane/pkg/prober/composite_prober_test.go [hunk 4]
```diff
@@ -111,7 +111,7 @@ func TestCompositeProber(t *testing.T) {
 			},
 			responseStatusCode:       http.StatusOK,
 			wantStatus:               StatusReady,
-			wantRequeueCountMin:      1,
+			wantRequeueCountMin:      0,
```
Comment (pierDipi) [inline control-plane/pkg/prober/composite_prober_test.go:114]: Why this change?

Comment (Cali0707) [reply control-plane/pkg/prober/composite_prober_test.go:114]: Oh that's my bad, I was messing around with some of the tests and I think I accidentally committed this, I'll revert it when I figure out how to fix the failing test!

Comment (pierDipi) [reply control-plane/pkg/prober/composite_prober_test.go:114]: maybe that's why is failing?

Comment (Cali0707) [reply control-plane/pkg/prober/composite_prober_test.go]: Looks like that fixed it locally...

**Comment by Cali0707** [comment]
/cc @pierDipi 
I think all the unit test issues should be fixed now so this should be ready to review

File: control-plane/pkg/prober/composite_prober_test.go [hunk 5]
```diff
@@ -224,8 +225,204 @@ func TestCompositeProber(t *testing.T) {
...
+				status := prober.Probe(ctx, tc.addressable, StatusReady)
+				return status == tc.wantStatus
+			}
+
+			require.Eventuallyf(t, probeFunc, 5*time.Second, 250*time.Millisecond, "")
+			require.Eventuallyf(t, func() bool { return wantHttpRequestCountMin.Load() == 0 }, 5*time.Second, 250*time.Millisecond, "got %d, want 0", wantHttpRequestCountMin.Load())
+			require.Eventuallyf(t, func() bool { return wantHttpsRequestCountMin.Load() == 0 }, 5*time.Second, 250*time.Millisecond, "got %d, want 0", wantHttpsRequestCountMin.Load())
+			// As we are expecting some of these tests to never resolve to Ready, the prober will potentially requeue more than the min number of times during the test period
+			// So, we want to test that we got at least as many requeues as the min we expect
+			require.Eventuallyf(t, func() bool { return requeueCountMin.Load() >= int64(tc.wantRequeueCountMin) }, 5*time.Second, 250*time.Millisecond, "got %d, want 0", requeueCountMin.Load())
```
Comment (pierDipi) [inline control-plane/pkg/prober/composite_prober_test.go:423]: Do we know why is this the case and does not happen in the other test?

Will we be able to detect that hard-looping doesn't happen? 

For example, currently, the test won't fail if `requeueCountMin` is like 1k (and so it's hard requeue-ing constantly)

Comment (Cali0707) [reply control-plane/pkg/prober/composite_prober_test.go:423]: Yeah, the other tests are testing expecting it to succeed. We are testing here with `exepectedStatus = StatusReady`, but we don't want it to actually succeed. So, it will keep retrying.

File: control-plane/pkg/prober/composite_prober.go [hunk 6]
```diff
@@ -41,6 +45,11 @@ func NewComposite(ctx context.Context, httpPort string, httpsPort string, IPsLis
 	}, nil
 }
 
+// NewCompositeNoTLS creates a composite prober which will fail if it attempts to probe an https address
+func NewCompositeNoTLS(ctx context.Context, httpPort string, IPsLister IPsLister, enqueue EnqueueFunc) (NewProber, error) {
+	return NewComposite(ctx, httpPort, "443", IPsLister, enqueue, &emptyCaCerts)
```
[the remark this answers is not in the document]
Comment (Cali0707) [reply control-plane/pkg/prober/composite_prober.go:50]: It's not super important what the port is here as this prober should never actually be used with a TLS address. So, I just used the canonical TLS/https port of 443

[the remark this answers is not in the document]
Comment (Cali0707) [reply control-plane/pkg/prober/composite_prober.go:50]: @Leo6Leo do you think there is a good reason to change this to 8443?

**Review by pierDipi — APPROVED** [review]
/lgtm
/approve
